### ISSUE TITLE
Parser: Should not treat assignment to group as pattern

### ISSUE THREAD

**Issue by pvdz** [babel/babel#8796]
## Bug Report

**Current Behavior**

When assigning to a group that is an assignment the parser currently creates a Pattern node for the group. It should consider this a syntax error.

**Input Code**

```js
(x = x) = x
```

**Expected behavior/code**

Syntax error.

**Babel Configuration (.babelrc, package.json, cli command)**

```js
        babelAst = babelParse(code, {
          sourceType: goalMode === GOAL_MODULE ? 'module' : 'script',
          strictMode: startInStrictMode ? true : false,
          plugins: ['estree'],
        }).program;
```

**Environment**
- Babel version(s): v7.0.0 (regular)
- Node/npm version: node 10
- OS: linux
- Monorepo only on sundays
- How you are using Babel: code

**Possible Solution**

Whatever the mechanism for checking assignable is in Babel should do so properly across parens. 

**Additional context/Screenshots**

Current:
```
{ type: 'Program',
  body:
   [ { type: 'ExpressionStatement',
       expression:
        { type: 'AssignmentExpression',
          left:
           { type: 'AssignmentPattern',
             left: { type: 'Identifier', name: 'x' },
             right: { type: 'Identifier', name: 'x' } },
          operator: '=',
          right: { type: 'Identifier', name: 'x' } } } ] }
```

Expected:

https://tc39.github.io/ecma262/#sec-assignment-operators-static-semantics-assignmenttargettype

This rule says that an assignment itself is not a valid assignment. 

Note that assignments are right associative, so `x = x = x` is really `x = (x = x)`, which is why this is still valid while still considering assignments invalid assignment targets. 

https://tc39.github.io/ecma262/#sec-semantics-static-semantics-assignmenttargettype

This rule says that the AssignmentTargetType state of a (non-arrow) group is the AssignmentTargetType of its contents.

In other words; it's an invalid AssignmentTargetType and it should throw a syntax error.

**Comment by nicolo-ribaudo** [comment]
Is that code actually valid?

**Comment by pvdz** [comment]
Now that you mention it, no. 

I double checked it and according to the spec `x = x` is not a valid AssignmentTargetType, which means you can't assign to it when wrapped in parens;

https://tc39.github.io/ecma262/#sec-assignment-operators-static-semantics-assignmenttargettype
> LeftHandSideExpression=AssignmentExpression is invalid assignment target
https://tc39.github.io/ecma262/#sec-semantics-static-semantics-assignmenttargettype
> PrimaryExpression:CoverParenthesizedExpressionAndArrowParameterList
> Let expr be CoveredParenthesizedExpression of CoverParenthesizedExpressionAndArrowParameterList.
> Return AssignmentTargetType of expr.

I'll update the ticket accordingly. And the bug in my parser which uncovered this problem in the first place. It was checking the rhs of an assignment for this state where it shouldn't even consider it at all. 

Thanks for keeping me on my toes :)

**Comment by nicolo-ribaudo** [comment]
If anyone wants to work on this bug, please leave a comment below :point_down: :point_down: :point_down: 

---

Because of JavaScript's syntactic complexity, it's not always possible to know if a given node is going to be an assignment target or an expression. For example, consider this code:
```js
[ { x: y = 3 } ] ...
```
If `...` is `;`, then it is an array literal containing an object literal. If it is `= foo`, then it is a destructuring pattern.

For this reason, we have a method to convert AST nodes parsed as expressions to assignment targets:

https://github.com/babel/babel/blob/1969e6b6aa7d90be3fbb3aca98ea96849656a55a/packages/babel-parser/src/parser/lval.js#L37-L118

As you can see, on line 74 we allow assigning to `AssignmentExpression`s, by converting them to `AssignmentPattern`s:
https://github.com/babel/babel/blob/1969e6b6aa7d90be3fbb3aca98ea96849656a55a/packages/babel-parser/src/parser/lval.js#L84-L88

This is needed in order to support default parameters and default values in destructuring:

```js
[ a = 2 ] = foo; // Ok!
```

In order to fix this bug, we probably need to add a `boolean` parameter to `toAssignable` to allow assignment patterns (e.g. `allowAssign`). It would only be set to `true` when the parent is an array/object pattern or a parameters list.

You can add a test for this bug in https://github.com/babel/babel/tree/master/packages/babel-parser/test/fixtures/core/categorized



---

If you don't know how to clone Babel, follow these steps: (you need to have `make` and `yarn` available on your machine).
1. Write a comment there to know other possible contributors that you are working on this bug.
1. Fork the repo
2. Run `git clone https://github.com/<YOUR_USERNAME>/babel.git && cd babel`
3. Run `yarn && make bootstrap`
4. Wait :hourglass_flowing_sand: 
5. Run `make watch` (or `make build` whenever you change a file)
6. Add a test (only `input.js`; `output.js` will be automatically generated)
7. Update the code!
8. `yarn jest [name-of-the-package-to-test]` to run the tests
    - If some test outputs don't match but the new results are correct, you can delete the bad `output.js` files and run the tests again
9. If it is working, run `make test` tu run all the tests
10. Run `git push` and open a PR!

**Comment by geordidearns** [comment]
@nicolo-ribaudo i’d be keen to try it out as a first PR to babel :)

**Comment by nicolo-ribaudo** [comment]
Go for it!
If you need help you can ask in our slack channel (https://slack.babeljs.io/)

**Comment by vivek12345** [comment]
@nicolo-ribaudo I see this issue has no update since 15 May. Can I pick this up?

**Comment by boweihan** [comment]
@nicolo-ribaudo is this currently being worked on / claimed? If not I'd love to give it a shot.

**Comment by nicolo-ribaudo** [comment]
It's yours :+1:

**Comment by hoalenh** [comment]
Thanks

[Nấm Linh Chi Đỏ Việt Nam](http://linhchinhathoang.com/san-pham.html), [Mua Nấm Linh Chi ](https://wordpress.com/view/namlinhchidonhatban.wordpress.com)
